File: tests/udi_tests/breakpoint.cxx
/*
 * Copyright (c) 2011-2015, UDI Contributors
 * All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <cstdlib>
#include <cstring>
#include <iostream>
#include <sstream>

#include "libudi.h"
#include "libuditest.h"
#include "test_bins.h"
#include "test_lib.h"

using std::cout;
using std::endl;
using std::stringstream;
using std::string;

class test_breakpoint : public UDITestCase {
    public:
        test_breakpoint()
            : UDITestCase(std::string("test_breakpoint")) {}
        virtual ~test_breakpoint() {}

        bool operator()(void);
};

static const char *TEST_BINARY = SIMPLE_BINARY_PATH;
static udi_address TEST_FUNCTION = SIMPLE_FUNCTION1;

static test_breakpoint testInstance;

bool test_breakpoint::operator()(void) {
    char *argv[] = { NULL };

    udi_proc_config config;
    config.root_dir = NULL;

    udi_error_e error_code;
    char *errmsg = NULL;
    udi_process *proc = create_process(TEST_BINARY, argv, NULL, &config, &error_code, &errmsg);
    free(errmsg);

    test_assert(proc != NULL);

    udi_thread *thr = get_initial_thread(proc);
    test_assert(thr != NULL);

    udi_error_e result = create_breakpoint(proc, TEST_FUNCTION);
    assert_no_error(proc, result);

    result = install_breakpoint(proc, TEST_FUNCTION);
    assert_no_error(proc, result);

    result = continue_process(proc);
    assert_no_error(proc, result);

    wait_for_breakpoint(thr, TEST_FUNCTION);

    udi_address pc;
    result = get_pc(thr, &pc);
    assert_no_error(proc, result);

    stringstream msg;
    msg << "Actual: "
        << std::hex
        << pc
        << std::dec
        << " Expected: "
        << std::hex
        << TEST_FUNCTION
        << std::dec;
    test_assert_msg(msg.str().c_str(), pc == TEST_FUNCTION);

    result = continue_process(proc);
    assert_no_error(proc, result);

    wait_for_exit(thr, EXIT_FAILURE);

    return true;
}

File: libudi/tests/breakpoint.rs
//
// Copyright (c) 2011-2017, UDI Contributors
// All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
#![deny(warnings)]

extern crate udi;

#[macro_use]
extern crate lazy_static;

mod native_file_tests;
mod utils;

use udi::Result;

#[test]
fn breakpoint() {
    if let Err(e) = breakpoint_test() {
        utils::print_error(e);
        panic!("breakpoint test failed");
    }
}

fn breakpoint_test() -> Result<()> {

    let addr = native_file_tests::get_test_metadata().simple_function1_addr();
    let exec_path = native_file_tests::get_test_metadata().simple_path().to_str().unwrap();

    let config = udi::ProcessConfig::new(None, utils::rt_lib_path());
    let argv = Vec::new();
    let envp = Vec::new();

    let proc_ref = udi::create_process(exec_path,
                                       &argv,
                                       &envp,
                                       &config)?;
    let thr_ref;
    {
        let mut process = proc_ref.lock()?;

        thr_ref = process.get_initial_thread();
        process.create_breakpoint(addr)?;
        process.install_breakpoint(addr)?;
        process.continue_process()?;
    }

    utils::wait_for_event(&proc_ref, &thr_ref, &udi::EventData::Breakpoint{ addr });

    let brkpt_addr = thr_ref.lock()?.get_pc()?;
    assert_eq!(addr, brkpt_addr);

    proc_ref.lock()?.continue_process()?;

    utils::wait_for_exit(&proc_ref, &thr_ref, 1);

    Ok(())
}
